### PR TITLE
Support attaching to an existing request queue in HTTP.SYS

### PR DESCRIPTION
**Pull request by bkatms**
Support attaching to an existing request queue in HTTP.SYS ( #5866).  This is an addendum to 
“https://github.com/aspnet/AspNetCore/issues/13461”.  

This is criticial (P0) functionality required by Office 365 in support of our new ASP.Net Core 3.X architecture for data access to Substrate Storage.  This is the storage that is mission critical for Office 365.

### CODE REVIEW

**Review by Tratcher — COMMENTED** [review]
Thanks, getting this rebased does help.

There are some open issues from #5866 that we hadn't figured out how to address yet. Most of these are centered around ensuring each listener on a queue is using the same configuration. For things like timeouts it's at most a warning that the config is out of sync, but for auth and PathBase it can affect how requests are processed. 
- PathBase will need a fallback mode when attached to existing queues where it can't look things up by index.
- For auth we might want to query the queue settings to see if they match and fail otherwise
- I'm not sure if we can query for the timeout settings.
- Some options can be changed after the server starts such as UrlPrefixes. We need to make sure those code paths are disabled (and throw?) if this instance did not create the queue.
- Do we need to surface an API that says if this instance did attach or create?

I'll pull your branch and see what I can work on in your off hours. Want to give me push access to your fork?

**Comment by anurse** [comment]
FYI @Tratcher check in with me before merging. As this is feature work we'll require special approval for 3.1 (shouldn't be a problem).

**Comment by Tratcher** [comment]
Didn't make any progress on this today other than squashing and relocating my old branch to make it easier to reference: https://github.com/aspnet/AspNetCore/commit/2c4824398786964af7bb0a0a470235946cab6de9

File: src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs
Comment (bkatms) [reply src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs]: In our case we will have an overall management process doing the main parent Q configuration and setting the settings on behalf of the ASP.Net Core processes as part of the integration into our routing fabric so I was worried less about that.  I can explain this more offline to you if you are interested in the full scenario in our our case.

We will not be changing UrlPrefixes at this time so we don't need that support and throwing is fine.

We don't care if we created ourselves or not,  HTTP.SYS will get rid of the Q when the last reference to it is gone.  However, I can see if you might want to return a flag.

Typically I assume that the process will not be able to attach to the named Q in the first place if it does not have permissions to do so. 

Thanks!

**Comment by Tratcher** [comment]
Made some progress, I pulled in the missing parts from the prior branch including tests and some misc fixes. I'll start on the actual review tomorrow.

File: src/Servers/HttpSys/ref/Microsoft.AspNetCore.Server.HttpSys.netcoreapp.cs
Comment (davidfowl) [reply src/Servers/HttpSys/ref/Microsoft.AspNetCore.Server.HttpSys.netcoreapp.cs]: For my edification, what’s controller?

Comment (Tratcher) [reply src/Servers/HttpSys/ref/Microsoft.AspNetCore.Server.HttpSys.netcoreapp.cs]: It's an orchestration role, the controller creates and configures the queues, but does not listen for requests. The customer already has one of these written in C++. We may consider exposing something like that eventually, but it would be separate from the web app model and wouldn't be an IServer.

File: src/Servers/HttpSys/src/RequestProcessing/Request.cs
Comment (Tratcher) [inline src/Servers/HttpSys/src/RequestProcessing/Request.cs]: Nevermind, this logic is wrong. We can only rely on UrlContext if this instance created the queue. Otherwise we don't know what those indexes were set to. We'll have to iterate through the collection and look for the longest match manually.

**Comment by Tratcher** [comment]
I think that's everything for now.
- I've dealt with PathBase. It can't query prefixes configured elsewhere but can do a match against ones you provide in the app.
- For auth you must make sure your schemes are in sync or it's not going to work well. We can discuss guardrails in the future if people have issues.
- Many settings are marked as not applicable in Attach mode.
- If using CreateOrAttach, there is a log statement indicating if it attached but there's no API to show the results in the app. This isn't a priority for 3.1 since our driving customer doesn't plan to use CreateOrAttach. In the future consider surfacing this as a server feature? I'm not sure what scenarios would require knowing the difference (HotAdd maybe?).
- Suppressed default urls when attached to an existing queue.

@halter73 @bkatms please review.

**Comment by anurse** [comment]
paging @davidfowl @rynowak for API review (not sure who else wants to be involved or if a GitHub team has been set up yet). 

The API here is super minimal, just [a new option that has an enum value](https://github.com/aspnet/AspNetCore/pull/14182/files#diff-3c0a65565dfe2587d01d4adebc2d2d02). LGTM.

File: src/Servers/HttpSys/src/RequestQueueMode.cs
Comment (davidfowl) [inline src/Servers/HttpSys/src/RequestQueueMode.cs]: Add a doc comment for the enum

File: src/Servers/HttpSys/src/UrlPrefixCollection.cs
Comment (davidfowl) [inline src/Servers/HttpSys/src/UrlPrefixCollection.cs]: Does this get called per request?

Comment (Tratcher) [reply src/Servers/HttpSys/src/UrlPrefixCollection.cs]: Yes, but the list should be short.

**Comment by bkatms** [comment]
The changes look fine to me.  Thank you.

**Comment by bkatms** [comment]
Note relative to "Created" - my understanding is that HTTP.SYS will automatically delete a queue if there are no active references to it by any process.

In our case we will have a management process create and configure the queues and pass then name into our ASP.NET Core process - this is similar to how IIS works when it starts W3WP.exe.  We will either directly spawn (security context inheritance) or set the ACLs appropriately - TBD.  In any case we plan on handling that part.

Let me know when you have a drop for me to begin testing on.

Cheers!

File: src/Servers/HttpSys/src/AuthenticationManager.cs
Comment (jkotalik) [inline src/Servers/HttpSys/src/AuthenticationManager.cs]: This should probably be a remark rather than summary.

File: src/Servers/HttpSys/src/HttpSysOptions.cs
Comment (jkotalik) [inline src/Servers/HttpSys/src/HttpSysOptions.cs]: Was this just a random find?

File: src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs
Comment (jkotalik) [inline src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs]: Shouldn't this technically check CreateOrAttach and OpenExisting?

File: src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs
Comment (jkotalik) [inline src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs]: Why was this added?

File: src/Servers/HttpSys/src/UrlPrefix.cs
Comment (jkotalik) [inline src/Servers/HttpSys/src/UrlPrefix.cs]: Get out of here with your dark magic.

File: src/Servers/HttpSys/src/UrlPrefix.cs
Comment (jkotalik) [inline src/Servers/HttpSys/src/UrlPrefix.cs]: To confirm, UrlPrefixes are created once, not once per request?

File: src/Servers/HttpSys/src/UrlPrefixCollection.cs
Comment (jkotalik) [inline src/Servers/HttpSys/src/UrlPrefixCollection.cs]: Add a comment with an example, especially for the longest match cases.

File: src/Servers/HttpSys/test/FunctionalTests/Listener/ServerOnExistingQueueTests.cs
Comment (jkotalik) [inline src/Servers/HttpSys/test/FunctionalTests/Listener/ServerOnExistingQueueTests.cs]: Add a test for longest match?

File: src/Servers/HttpSys/src/UrlPrefix.cs
Comment (Tratcher) [reply src/Servers/HttpSys/src/UrlPrefix.cs]: Right, these are server level config.

File: src/Servers/HttpSys/src/UrlPrefix.cs
Comment (Tratcher) [reply src/Servers/HttpSys/src/UrlPrefix.cs]: VS wrote that for me 😁

File: src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs
Comment (Tratcher) [reply src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs]: Eager clean up. If this secondary API call fails then the requestQueueHandle would only be cleaned up by its finalizer. The process is probably going to shut down anyways, but it's best to clean things up as locally as possible.

File: src/Servers/HttpSys/src/HttpSysOptions.cs
Comment (Tratcher) [reply src/Servers/HttpSys/src/HttpSysOptions.cs]: RequestQueueName was added as part of this feature, just in an earlier PR. And yeah, when writing tests I accidentally tried setting it to null and it threw null ref.

File: src/Servers/HttpSys/src/AuthenticationManager.cs
Comment (Tratcher) [reply src/Servers/HttpSys/src/AuthenticationManager.cs]: Remarks aren't visible in most situations.

File: src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs
Comment (Tratcher) [reply src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs]: That's why I'm checking the flags and not the enum.

**Comment by anurse** [comment]
@davidfowl this needs to be merged Very Soon. What if anything do you want to do for API review here?

File: src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs
Comment (jkotalik) [reply src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs]: > For CreateOrAttach that should only happen with a strange race condition.

Yeah, the strange race condition is what I was referring to.

**Comment by davidfowl** [comment]
> For auth you must make sure your schemes are in sync or it's not going to work well. We can discuss guardrails in the future if people have issues.

What’s the failure mode here?

**Comment by Tratcher** [comment]
> > For auth you must make sure your schemes are in sync or it's not going to work well. We can discuss guardrails in the future if people have issues.
> 
> > Many settings are marked as not applicable in Attach mode.
> 
> What’s the failure mode here?

The managed layer enforces AllowAnonymous and adds challenges based on what's configured. Worst case you end up requiring more auth than the underlying queue supports and the client gets rejected with 401s.

**Comment by bkatms** [comment]
I am good here.

Cheers, Brian
[he/him]

File: src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs
Comment (halter73) [reply src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs]: I don't think it's really worth guarding against that race condition, since to do so would require some sort of looping.

File: src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs
Comment (halter73) [inline src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs]: Should this log include the queue name?

**Review by halter73 — APPROVED** [review]

**Review by jkotalik — APPROVED** [review]

**Comment by Tratcher** [comment]
@anurse this is ready to merge.

**Pull request merged**
